homeassistant/components/mqtt/binary_sensor.py

```diff
@@ -114,6 +116,34 @@ def __init__(self, hass, config, config_entry, discovery_data):
 
         MqttEntity.__init__(self, hass, config, config_entry, discovery_data)
 
+    async def async_added_to_hass(self) -> None:
+        """Recover active entities with exire trigger."""
+        await super().async_added_to_hass()
+        if (
+            (expire_after := self._config.get(CONF_EXPIRE_AFTER)) is not None
+            and expire_after > 0
+            and (last_state := await self.async_get_last_state()) is not None
+            and last_state != STATE_UNAVAILABLE
+        ):
+            self._expired = False
+            self._state = last_state.state
+            expiration_at = dt_util.utcnow() + timedelta(seconds=expire_after)
```

Comment:
I think we should adjust `expiration_at` according to the age of the restored state, we can use `state.last_changed` for that

Reply:
done